### ISSUE TITLE
WebSocket: Define a communication protocol

### ISSUE THREAD

**Issue by JeanJPNM** [Sharlottes/MlogWatcher#6]
The creation of #4 made me think about how the communication method with this mod is going to become more complex as time goes on and more features are added. So I believe it may be necessary to use a communication protocol like json rpc to make the addition of new features more straightforward and backwards-compatible.

Of course, this is not required at the moment, but I think it may become relevant in the near future.

**Comment by JeanJPNM** [comment]
I can implement this change if you @Sharlottes and @cardillan agree.

**Comment by cardillan** [comment]
I think this is a very good idea. I wanted to propose adding support for sending schematics to Mlog Watcher via WebSockets (and retire [SchematicRefresher](https://github.com/cardillan/SchematicsRefresher) along the way) one day, but I don't have it all thought out yet.

Do you think you could provide a sample client code for Javascript? I currently have a Java integration (in the command line compiler) and Javascript integration in an unreleased version of Mindcode. I can handle Java, but I don't do front-end and sometimes struggle with the web app.

**Comment by JeanJPNM** [comment]
Sure, here is how the javascript code might look like. It uses a few adhoc constructs like the `JsonRpcConnection` class, but the most convenient (and less error prone) way to do it would be using a library like [json-rpc-2.0](https://www.npmjs.com/package/json-rpc-2.0) in this case.
```js

class JsonRpcConnection {
    constructor(socket) {
        // ids can be strings or numbers
        // this is just for demonstration purposes
        this.currentId = 0;
        // store callbacks for pending requests
        this.pending = new Map();
        this.socket = socket;
        // handle incoming messages
        socket.addEventListener("message", (event) => {
            const response = JSON.parse(event.data);
            if (response.id) {
                const resolve = this.pending.get(response.id);
                if (resolve) {
                    resolve(response);
                    this.pending.delete(response.id);
                }
            }
        });
    }
    send(method, params) {
        const id = this.currentId++;
        const message = JSON.stringify({
            jsonrpc: "2.0",
            id,
            method,
            params,
        });
        this.socket.send(message);
        return new Promise((resolve) => {
            // store the resolve function for later
            this.pending.set(id, resolve);
        });
    }
}
const socket = new WebSocket("ws://localhost:9992");
const conn = new JsonRpcConnection(socket);
// jsonrpc version of what we currently have
function sendSelectedProcessorUpdate(conn, code) {
    conn.send("updateSelectedProcessor", [code]);
}
// other possible methods we might want in the future
function sendProcessorUpdate(conn, position, code) {
    conn.send("updateProcessor", [position, code]);
}
// stub for demonstration purposes
async function sendListProcessorsRequest(conn) {
    const response = await conn.send("getProcessors", []);
    // TODO: handle errors
    if (response.error)
        return null;
    return response.result;
}
```

**Comment by Sharlottes** [comment]
> So I believe it may be necessary to use a communication protocol like [json rpc](https://www.jsonrpc.org/specification) to make the addition of new features more straightforward and backwards-compatible.

I've checked the specs, and I don't think we need to use that library, as it requires some boilerplate code for generality, but we're just “promising” and “validating” how JSON will be exchanged in our already implemented WS code. What's more, that library provides client and server classes, and since our client is in Mindustry mode with Java, we can't take full advantage of it.

Also, while it's reasonable to validate payloads with each other in anticipation of more complex features at this point, I think it's overkill to prepare for unplanned RPCs.

...so it looks like we only need to format the JSON to go into the message value of @JeanJPNM's `JsonRpcConnection#send` method, for example...

```ts
interface CustomPayload<T extends Record<string, unknown>> {
  id: string;
  data: T;
}
```

```js
// in JsonRpcConnection#send method
this.socket.send(JSON.stringify({
  id,
  // because it dosen't have to rpc, 'type' name will be naturely
  type: 'CHECK_STATUS', // number could be lighter
  data: { }
}))
```

**Comment by JeanJPNM** [comment]
Agreed, I think it's good enough for this purpose, so now we the only steps missing are:
- Document the how to connect with the mod (someone else might want to use this too)
- Document the valid request and response message types
- Actually implement it

**Comment by SinfullySoul** [comment]
any updated on this?

**Comment by cardillan** [comment]
I'd need to somehow resolve this. I'm updating MlogWatcher to also accept Base64-encoded schematics via websocket and add them to the schematics library (see #9), and I need a richer interface for that. Not sure what the current state of this issue is.

@JeanJPNM Would you be able to help?

**Comment by JeanJPNM** [comment]
Alright, so this is what I've thought about: Every remote procedure invocation must have the following shape:

```json
{
  "method": "<method name>",
  "method_version": 1,
  "invocation_id": 0,
  "params": {},
}
```

With there always being a response to let the client know about the status of the remote method invocation:
```json
{
  "status": "success or error, may add more in the future",
  "invocation_id": 0,
  "result": "it is up to the method to define, null if the method does not return anything"
}
```

With this, we can make updates the the API when necessary, bumping the version of each individual remote procedure.

Here is what the schemas would be for the existing/new methods:

```json
{
  "method": "update_selected_processor",
  "method_version": 1,
  "invocation_id": 0,
  "params": {
    "code": "base64 encoded logic text"
  }
}
```

```json
{
  "method": "update_processor_at",
  "method_version": 1,
  "invocation_id": 0,
  "params": {
    "x": 0,
    "y": 0,
    "code": "base64 encoded logic text"
  }
}
```

```json
{
  "method": "put_schematic_in_library",
  "method_version": 1,
  "invocation_id": 0,
  "params": {
    "schematic": "base64 encoded schematic text"
  }
}
```

**Comment by cardillan** [comment]
Thanks, this looks good! (I'd add an `overwrite` parameter to the `put_schematic_in_library` , with `yes` and `no`, but that's a detail.)

Whats the plan for the implementation? We need to add a json library, don't we? (Mindusty has something like json built in, but it is non-standard). Then perhaps some base class for implementing methods and we're good to go?

(Also, base64-encoding text means encoding the UTF-8 representation, right?)

**Comment by JeanJPNM** [comment]
That and make documentation for the protocol and the available methods.

**Comment by cardillan** [comment]
About the library, I was looking at https://github.com/ralfstx/minimal-json, because it is lightweight and has the same license as this project (MIT), but it is no longer maintained. Are there better options?

**Comment by JeanJPNM** [comment]
We could use https://github.com/FasterXML/jackson since we don't need to be very lightweight.

**Comment by cardillan** [comment]
No problem. Looks like we need to use the 2.x version, since 3.x requires Java 17 and Mindustry runs on Java 16.
